protocol/lavaprotocol/finalization_consensus.go

```diff
@@ -100,83 +99,113 @@ func (fc *FinalizationConsensus) insertProviderToConsensus(blockDistanceForFinal
+	defer func() {
+		fc.insertProviderToConsensus(latestBlock, blockDistanceForFinalizedData, finalizedBlocks, reply, relaySession, providerAddress)
+		fc.providerDataContainersMu.Unlock()
+	}()
 
-	if len(fc.currentProviderHashesConsensus) == 0 && len(fc.prevEpochProviderHashesConsensus) == 0 {
-		newHashConsensus := fc.newProviderHashesConsensus(blockDistanceForFinalizedData, providerAddress, latestBlock, finalizedBlocks, reply, req)
-		fc.currentProviderHashesConsensus = append(make([]ProviderHashesConsensus, 0), newHashConsensus)
+	var blockToHashesToAgreeingProviders BlockToHashesToAgreeingProviders
+	foundDiscrepancy, discrepancyBlock := fc.findDiscrepancy(finalizedBlocks, fc.currentBlockToHashesToAgreeingProviders)
+	if foundDiscrepancy {
+		blockToHashesToAgreeingProviders = fc.currentBlockToHashesToAgreeingProviders
 	} else {
-		inserted := false
-		// Looks for discrepancy with current epoch providers
-		// go over all consensus groups, if there is a mismatch add it as a consensus group and send a conflict
-		for _, consensus := range fc.currentProviderHashesConsensus {
-			err := fc.discrepancyChecker(finalizedBlocks, consensus)
-			if err != nil {
-				// TODO: bring the other data as proof
-				finalizationConflict = &conflicttypes.FinalizationConflict{RelayReply0: reply}
-				// we need to insert into a new consensus group before returning
-				// or create new consensus group if no consensus matched
-				continue
-			}
-
-			if !inserted {
-				// if no discrepency with this group and not inserted yet -> insert into consensus
-				fc.insertProviderToConsensus(blockDistanceForFinalizedData, &consensus, finalizedBlocks, latestBlock, reply, req, providerAddress)
-				inserted = true
-			}
-			// keep comparing with other groups, if there is a new message with a conflict we need to report it too
-		}
-		if !inserted {
-			// means there was a consensus mismatch with everything, so it wasn't inserted and we add it here
-			newHashConsensus := fc.newProviderHashesConsensus(blockDistanceForFinalizedData, providerAddress, latestBlock, finalizedBlocks, reply, req)
-			fc.currentProviderHashesConsensus = append(fc.currentProviderHashesConsensus, newHashConsensus)
-		}
-		if finalizationConflict != nil {
-			// means there was a conflict and we need to report
-			return finalizationConflict, utils.LavaFormatError("Simulation: Conflict found in discrepancyChecker", err)
+		// Could not find discrepancy, let's check with previous epoch
+		foundDiscrepancy, discrepancyBlock = fc.findDiscrepancy(finalizedBlocks, fc.prevEpochBlockToHashesToAgreeingProviders)
+		if foundDiscrepancy {
+			blockToHashesToAgreeingProviders = fc.prevEpochBlockToHashesToAgreeingProviders
+		} else {
+			// No discrepancy found, log and return nil
+			logSuccessUpdate()
+			return finalizationConflict, nil
 		}
+	}
+
+	replyFinalization := conflicttypes.NewRelayFinalizationFromRelaySessionAndRelayReply(relaySession, reply, consumerAddress)
```

Comment:
add an explanation here in what scenario we move on.  (didn't return) and what is going to happen.

Reply:
Can you explain what you meant by "add an explanation here in what scenario we move on"?